### PR TITLE
[BEAM-3153] Add Python processing-time timers and clock injection

### PR DESCRIPTION
**Pull request by mariapython**
- [x] Add processing-time timers
 - [x] Add clock to runners/direct
 - [x] Clock injection and propagation:
DirectRunner → EvaluationContext → WatermarkManager
 - [x] Use RealClock() for production and TestClock() for tests
 - [x] Add mechanism to detect not-yet fired realtime timers
 - [x] Refactoring: 
_TransformWatermarks’ extract_fired_timers → extract_transform_timers
WatermarkManager’s extract_fired_timers → extract_all_timers
EvaluationContext’s extract_fired_timers → extract_all_timers
 - [x] Add and improve comments
 - [ ] Tests

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by mariapython** [comment]
R: @charlesccychen

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by mariapython** [comment]
retest this please

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by mariapython** [comment]
Test timing out, will attempt a 3rd run. Perhaps related to https://issues.apache.org/jira/browse/BEAM-3040

**Comment by mariapython** [comment]
retest this please

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 1]
```diff
@@ -90,15 +92,14 @@ class DirectRunner(PipelineRunner):
   """Executes a single pipeline on the local machine."""
 
   # A list of PTransformOverride objects to be applied before running a pipeline
-  # using DirectRunner.
-  # Currently this only works for overrides where the input and output types do
-  # not change.
-  # For internal SDK use only. This should not be updated by Beam pipeline
-  # authors.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: Why the change here?

File: sdks/python/apache_beam/runners/direct/watermark_manager.py [hunk 2]
```diff
@@ -143,13 +143,18 @@ def _refresh_watermarks(self, applied_ptransform):
             for consumer in consumers:
               self._refresh_watermarks(consumer)
 
-  def extract_fired_timers(self):
+  def extract_all_timers(self):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/watermark_manager.py]: Can we use the previous method name?  This name is misleading because it isn't we still only extract fired timers.

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 3]
```diff
@@ -15,36 +15,35 @@
...
 
 class Clock(object):
-  """For internal use only; no backwards-compatibility guarantees."""
+  def current_time(self):
+    raise NotImplementedError()
 
-  def time(self):
-    """Returns the number of milliseconds since epoch."""
-    return int(time.time() * 1000)
+  def advance_time(self):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: Please add duration argument.

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 4]
```diff
@@ -15,36 +15,35 @@
...
+class RealClock(object):
+  def current_time(self):
+    return time.time()
 
-  Mock clock implementation for testing."""
 
-  def __init__(self, now_in_ms):
-    self._now_in_ms = now_in_ms
+class TestClock(object):
+  """Clock used for Testing"""
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: "testing".

Please document the unit used in `current_time` and `advance_by`.

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 5]
```diff
@@ -15,36 +15,35 @@
...
+For internal use only. No backwards compatibility guarantees.
+"""
 from __future__ import absolute_import
 
 import time
 
 
 class Clock(object):
-  """For internal use only; no backwards-compatibility guarantees."""
+  def current_time(self):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: Can we rename this `time()` to be consistent with the way `time.time()` is called?

File: sdks/python/apache_beam/runners/direct/executor.py [hunk 6]
```diff
@@ -545,22 +545,19 @@ def call(self):
           self._executor.executor_service.submit(self)
 
     def _should_shutdown(self):
-      """_should_shutdown checks whether pipeline is completed or not.
+      """Checks whether the pipeline is completed and should be shut down.
 
-      It will check for successful completion by checking the watermarks of all
-      transforms. If they all reached the maximum watermark it means that
-      pipeline successfully reached to completion.
+      If there is anything in the queue of tasks to do, do not shut down.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/executor.py]: Are we checking whether there are pending processing time timers?

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 7]
```diff
@@ -90,15 +92,14 @@ class DirectRunner(PipelineRunner):
...
+  # using DirectRunner. Currently, this only works for overrides where the input
+  # and output types do not change.
+  # For internal use only; no backwards-compatibility guarantees.
   _PTRANSFORM_OVERRIDES = []
 
   def __init__(self):
     self._cache = None
+    self._use_test_clock = False  # use RealClock() in production
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: Can you document that this is used for testing when TestStream is used in the pipeline?

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 1 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: > **charlesccychen** wrote:
> Why the change here?


At the time of the 2.0 release we tried to normalize the message to "For internal use only; no backwards-compatibility guarantees." (https://issues.apache.org/jira/browse/BEAM-2241)

File: sdks/python/apache_beam/runners/direct/watermark_manager.py (hunk 2 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/watermark_manager.py]: > **charlesccychen** wrote:
> Can we use the previous method name?  This name is misleading because it isn't we still only extract fired timers.


I meant to do the following changes:
WatermarkManager's extract_fired_timers --> extract_all_timers (meaning for all transforms)
_TransformWatermarks' extract_fired_timers --> extract_transform_timers (meaning per transform).
I meant to accomplish two things: 1) differentiate between "for all" and "per transfer" and 2) we are no longer just extracting the fired timers. I understand 2) is not crucial, as the other return is basically a flag.
Another (verbose) option is:  extract_fired_timers and extract_fired_timers_per_transform.

File: sdks/python/apache_beam/runners/direct/executor.py (hunk 6 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/executor.py]: > **charlesccychen** wrote:
> Are we checking whether there are pending processing time timers?


Yes, and I consider that a "task to do," which is looked at in _is_executing() and that is where I added a comment about timers being set.

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 4 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> "testing".
> 
> Please document the unit used in `current_time` and `advance_by`.


Done.

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 3 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> Please add duration argument.


Done.

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 5 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> Can we rename this `time()` to be consistent with the way `time.time()` is called?


Done.

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 7 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: > **charlesccychen** wrote:
> Can you document that this is used for testing when TestStream is used in the pipeline?


I mention that a few lines below, when I turn on the flag.

**Comment by mariapython** [comment]
retest this please

**Comment by aaltay** [comment]
What is the next action on this PR? @charlesccychen could you please make another pass?

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 8]
```diff
@@ -15,36 +15,37 @@
...
+  def advance_time(self, advance_by):
+    """Advances the clock by a number of miliseconds."""
+    raise NotImplementedError()
 
 
-class MockClock(Clock):
-  """For internal use only; no backwards-compatibility guarantees.
+class RealClock(object):
+  def time(self):
+    return int(time.time() * 1000)
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: Can you please change this back to seconds?

File: sdks/python/apache_beam/transforms/trigger.py [hunk 9]
```diff
@@ -1138,20 +1138,36 @@ def clear_state(self, window, tag):
     if not self.state[window]:
       self.state.pop(window, None)
 
-  def get_timers(self, clear=False, watermark=MAX_TIMESTAMP):
+  def get_timers(self, clear=False, watermark=MAX_TIMESTAMP, current_time=None):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger.py]: Let's change this argument name to "processing_time".

File: sdks/python/apache_beam/transforms/trigger.py [hunk 10]
```diff
@@ -1138,20 +1138,36 @@ def clear_state(self, window, tag):
...
+
+    Expiration is measured against the watermark for event-time timers,
+    and against a wall clock for processing-time timers.
+    """
     expired = []
+    has_realtime_timer = False
     for window, timers in list(self.timers.items()):
       for (name, time_domain), timestamp in list(timers.items()):
-        if timestamp <= watermark:
+        if time_domain == 'REAL_TIME':
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger.py]: Please use the `TimeDomain` enum (i.e. `TimeDomain.WATERMARK`, etc) here.

File: sdks/python/apache_beam/transforms/trigger.py [hunk 11]
```diff
@@ -1138,20 +1138,36 @@ def clear_state(self, window, tag):
     if not self.state[window]:
       self.state.pop(window, None)
 
-  def get_timers(self, clear=False, watermark=MAX_TIMESTAMP):
+  def get_timers(self, clear=False, watermark=MAX_TIMESTAMP, current_time=None):
+    """Gets expired timers and reports if there
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger.py]: Let's try to fit in one line:

```python
"""Returns expired timers and existence of REAL_TIME timers."""
```

**Comment by mariapython** [comment]
retest this please

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 8 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> Can you please change this back to seconds?


Done.

File: sdks/python/apache_beam/transforms/trigger.py (hunk 9 above)
Comment (mariapython) [inline sdks/python/apache_beam/transforms/trigger.py]: > **charlesccychen** wrote:
> Let's change this argument name to "processing_time".


Done.

File: sdks/python/apache_beam/transforms/trigger.py (hunk 10 above)
Comment (mariapython) [inline sdks/python/apache_beam/transforms/trigger.py]: > **charlesccychen** wrote:
> Please use the `TimeDomain` enum (i.e. `TimeDomain.WATERMARK`, etc) here.


Done.

File: sdks/python/apache_beam/transforms/trigger.py (hunk 11 above)
Comment (mariapython) [inline sdks/python/apache_beam/transforms/trigger.py]: > **charlesccychen** wrote:
> Let's try to fit in one line:
> 
> ```python
> """Returns expired timers and existence of REAL_TIME timers."""
> ```


Yes, it can be done for the first sentence, but I think the second is more important (explaining the addition of 'time_marker'). When I added the docstring initially, I considered writing that second sentence by the 'if' but I found it more clunky and, since the expiration is mentioned in the first sentence, referring to it in the second flowed better.

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 12]
```diff
@@ -15,36 +15,37 @@
...
 class Clock(object):
-  """For internal use only; no backwards-compatibility guarantees."""
-
   def time(self):
     """Returns the number of milliseconds since epoch."""
-    return int(time.time() * 1000)
+    raise NotImplementedError()
+
+  def advance_time(self, advance_by):
+    """Advances the clock by a number of miliseconds."""
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: Please fix comment to indicate seconds.

File: sdks/python/apache_beam/runners/direct/clock.py [hunk 13]
```diff
@@ -15,36 +15,37 @@
...
 from __future__ import absolute_import
 
 import time
 
 
 class Clock(object):
-  """For internal use only; no backwards-compatibility guarantees."""
-
   def time(self):
     """Returns the number of milliseconds since epoch."""
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/clock.py]: Please fix comment to indicate seconds.

File: sdks/python/apache_beam/runners/direct/executor.py [hunk 14]
```diff
@@ -602,7 +599,17 @@ def _fire_timers(self):
...
-      """Returns True if there is at least one non-blocked TransformExecutor."""
+      """Checks whether the job is still executing.
+
+      Returns:
+        True if there are any timers set or if there is at least
+        one non-blocked TransformExecutor active."""
+
+      watermark_manager = self._executor.evaluation_context._watermark_manager
+      _, any_unfired_timers = watermark_manager.extract_all_timers()
+      if any_unfired_timers:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/executor.py]: Can you more clearly indicate (by naming or comment) that this is for unfired real time timers?

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 12 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> Please fix comment to indicate seconds.


Done.

File: sdks/python/apache_beam/runners/direct/clock.py (hunk 13 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/clock.py]: > **charlesccychen** wrote:
> Please fix comment to indicate seconds.


Done.

File: sdks/python/apache_beam/runners/direct/executor.py (hunk 14 above)
Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/executor.py]: > **charlesccychen** wrote:
> Can you more clearly indicate (by naming or comment) that this is for unfired real time timers?


Done.

**Comment by chamikaramj** [comment]
Thanks. Merging.
